### PR TITLE
refactor avro reader and writer

### PR DESCRIPTION
**Pull request by chenjunjiedada**
Currently, Iceberg has two Avro reader/writer implementations, one is for internal use and another is for public use. This PR moves the public Avro reader and writer to iceberg-data subproject so that developers can avoid using the internal one which will be removed from public API. This PR also rename the reader and writer with pattern {`file format`} + {`data model`} + reader/writer, as discussed in #1152.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/data/avro/DataReader.java [hunk 1]
```diff
@@ -19,192 +19,25 @@
...
+ * @deprecated will be removed, please use AvroGenericReader instead.
+ */
+@Deprecated
+public class DataReader<T> extends BaseAvroGenericReader<T> {
+  DataReader(Schema expectedSchema, org.apache.avro.Schema readSchema, Map<Integer, ?> idToConstant) {
+    super(expectedSchema, readSchema, idToConstant);
+  }
 
-  public static <D> DataReader<D> create(org.apache.iceberg.Schema expectedSchema, Schema readSchema) {
+  public static <D> DataReader<D> create(Schema expectedSchema, org.apache.avro.Schema readSchema) {
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/data/avro/DataReader.java:35]: I think this changes the semantics of deprecated `DataReader#create`,  the arguments has been swapped.

Comment (chenjunjiedada) [reply core/src/main/java/org/apache/iceberg/data/avro/DataReader.java:35]: No, I import iceberg schema instead of avro schema. Let me update this to make it clear.

**Comment by rdblue** [comment]
@chenjunjiedada, can you summarize the changes here in the PR description?

Looks like this is primarily a rename, `DataReader` -> `AvroGenericReader`, using a base class, `BaseAvroGenericReader` for the read path. That name seems confusing to me because we now have `o.a.i.avro.GenericAvroReader` and `o.a.i.data.avro.AvroGenericReader`.

I understand the goal of removing the reader for Avro generics. What is your plan to get there?

**Comment by chenjunjiedada** [comment]
@rdblue, just updated PR description. The first step for refactoring I 'd like to take is to move public Avro reader and writer to iceberg-data subproject so that users can avoid referring the reader/writer in iceberg-core, and then remove `DataReader` and `DataWriter` in one of the later release. It does look confusing between `GenericAvroReader` and `AvroGenericReader`, how about we rename the internal one to `AvroReader`?

**Comment by rdblue** [comment]
I don't think that these renames are a good idea right now. There isn't a big difference between having `AvroReader` and `GenericAvroReader` or `DataReader` and `AvroGenericReader`. I think the rename makes it _more_ confusing, but even if it is slightly better, I don't think it is worth the code churn.

I think a better path to removing the support for Avro generics is to wait until the end to rename and start by moving internal classes to use a `ValueReader` tree that doesn't depend on Avro's object model.

**Comment by chenjunjiedada** [comment]
OK, make sense to me. Let me think about replacing `IndexedRecord` with `Record` at first.

**Comment by chenjunjiedada** [comment]
The PR: https://github.com/apache/iceberg/pull/1219 starting to replace `IndexedRecord` to `Record`.
